parameters.md
# Active Parameters
 For numerical parameters, the range of choices are abreviated as follows:
- '-' : the range includes negative numbers.
- '0' : the range includes zero.
- '+' : the range includes positive numbers.

| Parameter                             | Choices                                                                        | Default                                   | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                          | Type         | Unit Support   | Deprecated   | Hint   |
|:--------------------------------------|:-------------------------------------------------------------------------------|:------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------|:---------------|:-------------|:-------|
| AccretionFeedback            | yes, no                                                                        | DiskFeedback [DiskFeedback default = no]                                     | Let planets accrete without disk feedback being enabled.                                                                                                                                                                                                                                                                                                                                                                                                             | bool         | False          |              |        |
| AdiabaticIndex                        | +                                                                              | 1.4                                       | Numerical value or FIT_ISOTHERMAL (only for polytropic equation of state)                                                                                                                                                                                                                                                                                                                                                                                            | double       | False          |              |        |
| AlphaCold                             | 0+                                                                             | 0.01                                      | Alpha value for the cold state in scurve alpha.                                                                                                                                                                                                                                                                                                                                                                                                                      | double       | False          |              |        |
| AlphaHot                              | 0+                                                                             | 0.1                                       | Alpha value for the hot state in scurve alpha.                                                                                                                                                                                                                                                                                                                                                                                                                       | double       | False          |              |        |
| AlphaMode                             | 0, 1, 2, 3                                                                     | 0                                         | Select how the viscosity is calculated. 0: Const alpha  1: Scurve Alpha  2: Nbody distance dependend  3: Scurve Alpha with ion fraction                                                                                                                                                                                                                                                                                                                              | int          | False          |              |        |
| ArtificialViscosity                   | none, TW, SN                                                                   | SN                                        | Type of artificial viscosity [default = SN] TODO: change default to TW                                                                                                                                                                                                                                                                                                                                                                                               | string       | False          |              |        |
| ArtificialViscosityDissipation        | yes, no                                                                        | True                                      | Use artificial viscosity in dissipation function.                                                                                                                                                                                                                                                                                                                                                                                                                    | bool         | False          |              |        |
| ArtificialViscosityFactor             | +                                                                              | 1.41                                      | Artificial viscosity factor/constant (von Neumann-Richtmyer constant).                                                                                                                                                                                                                                                                                                                                                                                               | double       | False          |              |        |
| AspectRatio                           | +                                                                              | 0.05                                      | Set the reference value for the aspect ratio: h = ASPECTRATIO * r^{1+flaringindex}                                                                                                                                                                                                                                                                                                                                                                                   | double       | False          |              |        |
| AspectRatioMode                       | 0, 1, 2                                                                        | 0                                         | Compute aspectratio with respect to: 0: Primary object, 1: Nbody system, 2: Nbody center of mass                                                                                                                                                                                                                                                                                                                                                                     | int          | False          |              |        |
| BitwiseExactRestarting                | yes, no                                                                        | False                                     | Write out Qplus and Qminus 2D arrays to snapshot for bitwise exact restarting. This is really usefull to test development changes in the code.                                                                                                                                                                                                                                                                                                                       | bool         | False          |              |        |
| BodyForceFromPotential                | yes, no                                                                        | True                                      | Consider the gravitational force from the Nbody onto the disk via the potential.                                                                                                                                                                                                                                                                                                                                                                                     | bool         | False          |              |        |
| CFL                                   | +                                                                              | 0.5                                       | CFL factor.                                                                                                                                                                                                                                                                                                                                                                                                                                                          | double       | False          |              |        |
| CFLmaxVar                             | +                                                                              | 1.1                                       | Maximum variation of the CFL timestep over one iteration.                                                                                                                                                                                                                                                                                                                                                                                                            | double       | False          |              |        |
| CICPLANET                             | yes, no                                                                        | False                                     | Initialize planets in center of cell. Only works for zero eccentricity.                                                                                                                                                                                                                                                                                                                                                                                              | bool         | False          |              |        |
| CartesianParticles                    | yes, no                                                                        | False                                     | Particle coordinates are stored in memory and in output files in cartesian coordinates.                                                                                                                                                                                                                                                                                                                                                                              | bool         | False          |              |        |
| CenterProfileDensityCorrectionFactor  |                                                                                | 1                                         | ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | double       | False          |              |        |
| CircumBinaryDecayExponent             |                                                                                | 0.75                                      | ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | double       | False          |              |        |

ChangeLog.md
# Changelog of FargoCPT

Changes are in inverse chronological order, latest changes come first.

Prior to version 1.3.0, there was no versioning, so we used the latest output version (1.3) and made it the latest code version.

Syntax of the versions are: major, minor, revision
- major: major restructuring of code or config files or output
- minor: new features, new options, new config file options, renamed parameters, new or changed single output files
- revision: bug fixes

## Version 1.4
- rename 'vtheta' and 'vphi' both to 'vazi'. The output file of azimuthal velocity is now called 'vazi.dat' instead of 'vtheta.dat'. You might have to rename files to restart data.
- executable file renamed from `fargocpt` to `fargocpt_exe` to allow python cli to be called `fargocpt`
- added installable python package (`pip install <repo path>/python_module`) which installs a command line interface called `fargocpt`. Start a simulation with `fargocpt run` and check the data with `fargocpt data <output dir>`
- `planet{n}.dat/bin` files renamed to `nbody{n-1}.dat/bin`, thus `nbody0.dat` instead of `planet1.dat`. This requires renaming files to restart older data.
- nbody indices start at 0 now instead of at 1.
- the `run_fargo` file in the repository root is now a python script that uses the binary from the `bin` folder

## Version 1.3
- the code now has a version string, run ./bin/fargocpt --version
- there is now a `parameters.md` file with a table of all parameters.- changed the name of some parameters including Ntot->Nsnapshot, Ninterm->Nmonitor, DT->MonitorTimestep, run ./Tools/param_names/replace_parameter_names.py on a yml config file to update the names.
- code now exits in case of unknown parameters in the setup file. This help spotting typos.
- info about particle data is now in the `infoParticles.yml` file
- info about 1D output is now in the `info1D.yml` file
- copies of log written to stdout are written to a output and error file in the log directory in the output directory
- backtraces are written to a separate file for each process
- constants file in yaml containing all physical constants in the code
- units file in yaml containing all units in the code including code unit value in cgs and the cgs unit symbols
- Radiative Diffusion Tolerance parameter is now an absolute quantity, if only a number is supplied, its multiplied by the Temperature unit, otherwise its parsed.
- there is a file `info2D.yml` which describes the 2D output variables including their array size and units
- self gravity acceleration can now be written out to the snapshots using `WriteSGAccelRad` and `WriteSGAccelAzi` parameters
- Beta cooling includes a new `floor` mode in which the temperature is relaxed to the floor value. Can be configured using the `CoolingBetaReference` parameter which can be `zero` (to zero Kelvin), `reference` (to the reference/damping values), `model` (to the temperature from aspect ratio). This will break your old config.
- Cooling from the disk surface: the config parameter `CoolingRadiativeLocal` has been replaced by `SurfaceCooling` which can take the value `no|false|off`, `thermal` (the existing mode), or `scurve` (new mode).
- there is a file `info2D.yml` which describes the 2D output variables including their array size and units
- self gravity acceleration can now be written out to the snapshots using `WriteSGAccelRad` and `WriteSGAccelAzi` parameter
- bigplanet.dat files are now called planet.dat files, now fargocpt output v1.2
- restructured the output directory into snapshots and monitor directories. Each snapshot dir contains all information to restart the simulation.
- replaced the default executable to run, now run_fargo starts the simulations and calls bin/fargocpt and handles the parallel config
- simplify build process and remove FARGO_ARCH, use specific env variables now (see arch.defs)
- backtraces are written to file by default
- irradiation is now controlled by setting a temperature > 0 in the nbody config rather than the "HeatingStar" flag
- removed default star, central object needs to be declared as a planet now
- OUTPUTDIR is now in output namespace and called outdir
- physical constants (gravitational constant, kB, atomic mass unit, Planck constant) from LLNL units library which follows the NIST constants and the 2019 redefinition of SI
- stellar heating rampup time in units of T0 rather than DT
- beta cooling rampup time in units of T0
- thinknesssmoothing factor 0.6 by default (was 0.0)
- Changed tau_factor default value from 1 to 0.5.


src/boundary_conditions/config.cpp
/**
	\file boundary_conditions.cpp

	All boundary conditions are handled in this file.
*/
#ifdef _OPENMP
#include <omp.h>
#endif

#include "boundary_conditions.h"

#include "../config.h"
#include "../logging.h"
#include "../units.h"
#include "../LowTasks.h"
#include "../global.h" // For RMAX

#include <string>

namespace boundary_conditions
{

void (*sigma_inner_func)(t_polargrid &, t_polargrid &, t_data &);
void (*sigma_outer_func)(t_polargrid &, t_polargrid &, t_data &);
void (*energy_inner_func)(t_polargrid &, t_polargrid &, t_data &);
void (*energy_outer_func)(t_polargrid &, t_polargrid &, t_data &);
void (*vrad_inner_func)(t_polargrid &, t_polargrid &, t_data &);
void (*vrad_outer_func)(t_polargrid &, t_polargrid &, t_data &);
void (*vaz_inner_func)(t_polargrid &, t_polargrid &, t_data &);
void (*vaz_outer_func)(t_polargrid &, t_polargrid &, t_data &);
void (*special_inner_func)(t_polargrid &, t_polargrid &, t_data &);
void (*special_outer_func)(t_polargrid &, t_polargrid &, t_data &);

std::string sigma_inner_name = "";
std::string sigma_outer_name = "";
std::string energy_inner_name = "";
std::string energy_outer_name = "";
std::string vrad_inner_name = "";
std::string vrad_outer_name = "";
std::string vaz_inner_name = "";
std::string vaz_outer_name = "";
std::string composite_inner_name = "";
std::string composite_outer_name = "";

double keplerian_azimuthal_outer_factor = 1.0;
double keplerian_azimuthal_inner_factor = 1.0;
double keplerian_radial_outer_factor = 1.0;
double keplerian_radial_inner_factor = 1.0;


bool domegadr_zero;

double viscous_outflow_speed;
bool rochelobe_overflow;
unsigned int rof_planet;
double rof_temperature;
double rof_mdot;
double rof_rampingtime;
bool rof_variableTransfer;
double rof_averaging_time;
double rof_gamma;


/* 
 * Dummy to call when all variables are handled by one special function.
*/



/*
 * Get the type of an individual variable's boundary condition for a given key.
 * This function takes into account types set by a composite boundary condition.
 * See function for composites below.
 */
static std::string get_type(const std::string key, std::string &name)
{
    std::string str = config::cfg.get_lowercase(key, "infer");
    if (str == "infer") {
	if (name == "") { // nothing to infer here
	    throw std::runtime_error(
		"Can not infer '" + key +
		"' when 'InnerBoundary/OuterBoundary' is set to 'individual'");
	} else {
	    return name;
	}
    } else { // set individual boundary directly
	if (name != "") {
	    logging::print_master(LOG_INFO "BC: Overwriting %s from %s set via composite to %s.\n",
			  key.c_str(), name.c_str(), str.c_str());
	}
	name = str;
	return str;
    }
}

static void sigma_inner()
{
    std::string str = get_type("InnerBoundarySigma", sigma_inner_name);

    if (str == "zerogradient") {
	sigma_inner_func = zero_gradient_inner;
    } else if (str == "diskmodel") {
	sigma_inner_func = diskmodel_inner_sigma;
    } else if (str == "reference") {
	sigma_inner_func = reference_inner;
    } else if (str == "none") {
	sigma_inner_func = no_operation;
    if (composite_inner_name == "") {
        throw std::runtime_error(
            "Can not set InnerBoundarySigma to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error(
	    "Unknown boundary condition for sigma inner: " + str);
    }

    logging::print_master(LOG_INFO "BC: Sigma inner = %s\n",
			  sigma_inner_name.c_str());
}

static void sigma_outer()
{
    const std::string str = get_type("OuterBoundarySigma", sigma_outer_name);

    if (str == "zerogradient") {
	sigma_outer_func = zero_gradient_outer;
    } else if (str == "diskmodel") {
	sigma_outer_func = diskmodel_outer_sigma;
    } else if (str == "reference") {
	sigma_outer_func = reference_outer;
    } else if (str == "none") {
    sigma_outer_func = no_operation;
    if (composite_outer_name == "") {
        throw std::runtime_error(
            "Can not set OuterBoundarySigma to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error(
	    "Unknown boundary condition for sigma outer: " + str);
    }
    logging::print_master(LOG_INFO "BC: Sigma outer = %s\n",
			  sigma_outer_name.c_str());
}

static void energy_inner()
{
    const std::string str = get_type("InnerBoundaryEnergy", energy_outer_name);

    if (str == "zerogradient") {
	energy_inner_func = zero_gradient_inner;
    } else if (str == "diskmodel") {
	energy_inner_func = diskmodel_inner_energy;
    } else if (str == "reference") {
	energy_inner_func = reference_inner;
    } else if (str == "none") {
    energy_inner_func = no_operation;
    if (composite_inner_name == "") {
        throw std::runtime_error(
            "Can not set InnerBoundaryEnergy to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error(
	    "Unknown boundary condition for energy inner: " + str);
    }
    logging::print_master(LOG_INFO "BC: Energy inner = %s\n",
			  energy_inner_name.c_str());
}

static void energy_outer()
{
    const std::string str = get_type("OuterBoundaryEnergy", energy_outer_name);

    if (str == "zerogradient") {
	energy_outer_func = zero_gradient_outer;
    } else if (str == "diskmodel") {
	energy_outer_func = diskmodel_outer_energy;
    } else if (str == "reference") {
	energy_outer_func = reference_outer;
    } else if (str == "none") {
    energy_outer_func = no_operation;
    if (composite_outer_name == "") {
        throw std::runtime_error(
            "Can not set OuterBoundaryEnergy to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error(
	    "Unknown boundary condition for energy outer: " + str);
    }
    logging::print_master(LOG_INFO "BC: Energy outer = %s\n",
			  energy_outer_name.c_str());
}

static void vrad_inner()
{
    const std::string str = get_type("InnerBoundaryVrad", vrad_inner_name);

    if (str == "zerogradient") {
	vrad_inner_func = zero_gradient_inner;
    } else if (str == "reference") {
	vrad_inner_func = reference_inner;
    } else if (str == "reflecting") {
	vrad_inner_func = reflecting_inner;
    } else if (str == "outflow") {
	vrad_inner_func = outflow_inner;
    } else if (str == "viscous") {
	vrad_inner_func = viscous_outflow_inner;
    } else if (str == "keplerian") {
	vrad_inner_func = keplerian_radial_inner;
    } else if (str == "none") {
    vrad_inner_func = no_operation;
    if (composite_inner_name == "") {
        throw std::runtime_error(
            "Can not set InnerBoundaryVrad to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error("Unknown boundary condition for vrad inner: " +
				 str);
    }
    // parse parameters
    keplerian_radial_inner_factor =
	config::cfg.get<double>("InnerBoundaryVradKeplerianFactor", 0.1);
    // log
    logging::print_master(LOG_INFO "BC: Vrad inner = %s\n",
			  vrad_inner_name.c_str());
}

static void vrad_outer()
{
    const std::string str = get_type("OuterBoundaryVrad", vrad_outer_name);

    if (str == "zerogradient") {
	vrad_outer_func = zero_gradient_outer;
    } else if (str == "reference") {
	vrad_outer_func = reference_outer;
    } else if (str == "reflecting") {
	vrad_outer_func = reflecting_outer;
    } else if (str == "outflow") {
	vrad_outer_func = outflow_outer;
    } else if (str == "viscous") {
	vrad_outer_func = viscous_inflow_outer;
    } else if (str == "keplerian") {
	vrad_outer_func = keplerian_radial_outer;
    } else if (str == "none") {
    vrad_outer_func = no_operation;
    if (composite_outer_name == "") {
        throw std::runtime_error(
            "Can not set OuterBoundaryVrad to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error("Unknown boundary condition for vrad outer: " +
				 str);
    }
    // parse parameters
    keplerian_radial_outer_factor =
	config::cfg.get<double>("OuterBoundaryVradKeplerianFactor", 0.1);
    // log
    logging::print_master(LOG_INFO "BC: Vrad outer = %s\n",
			  vrad_outer_name.c_str());
}

static void vaz_inner()
{
    std::string defval = "keplerian";
    if (vaz_inner_name == "none") { 
        defval = "none";
    }

    const std::string str = config::cfg.get_lowercase("InnerBoundaryVazi", defval);
    if (vaz_inner_name == "none" && str != "none") {
        logging::print_master(LOG_WARNING "Your choice '%s' for InnerBoundaryVazi might have no effect because it might be overwritten by a composite boundary condition.\n", str.c_str());

    }

    vaz_inner_name = str;
    if (str == "zerogradient") {
	vaz_inner_func = zero_gradient_inner;
    } else if (str == "reference") {
	vaz_inner_func = reference_inner;
    } else if (str == "zeroshear") {
	vaz_inner_func = zero_shear_inner;
    } else if (str == "balanced") {
	vaz_inner_func = balanced_inner;
    } else if (str == "keplerian") {
	vaz_inner_func = keplerian_azimuthal_inner;
    } else if (str == "none") {
    vaz_inner_func = no_operation;
    if (composite_inner_name == "") {
        throw std::runtime_error(
            "Can not set InnerBoundaryVazi to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error("Unknown boundary condition for vaz inner: " +
				 str);
    }
    // parse parameters
    keplerian_azimuthal_inner_factor =
	config::cfg.get<double>("InnerBoundaryVaziKeplerianFactor", 1.0);
    // log
    logging::print_master(LOG_INFO "BC: Vaz inner = %s\n",
			  vaz_inner_name.c_str());
}

static void vaz_outer()
{
    std::string defval = "keplerian";
    if (vaz_outer_name == "none") { 
        defval = "none";
    }
    const std::string str = config::cfg.get_lowercase("OuterBoundaryVazi", defval);
    if (vaz_outer_name == "none" && str != "none") {
        logging::print_master(LOG_WARNING "Your choice '%s' for OuterBoundaryVazi might have no effect because it might be overwritten by a composite boundary condition.\n", str.c_str());

    }

    vaz_outer_name = str;
    if (str == "zerogradient") {
	vaz_outer_func = zero_gradient_outer;
    } else if (str == "reference") {
	vaz_outer_func = reference_outer;
    } else if (str == "zeroshear") {
	vaz_outer_func = zero_shear_outer;
    } else if (str == "balanced") {
	vaz_outer_func = balanced_outer;
    } else if (str == "keplerian") {
	vaz_outer_func = keplerian_azimuthal_outer;
    } else if (str == "none") {
    vaz_outer_func = no_operation;
    if (composite_outer_name == "") {
        throw std::runtime_error(
            "Can not set OuterBoundaryVazi to 'none' without selecting special boundary (e.g. 'custom').");
    }
    } else {
	throw std::runtime_error("Unknown boundary condition for vaz outer: " +
				 str);
    }
    // parse parameters
    keplerian_azimuthal_outer_factor =
	config::cfg.get<double>("OuterBoundaryVaziKeplerianFactor", 1.0);
    // log
    logging::print_master(LOG_INFO "BC: Vaz outer = %s\n",
			  vaz_outer_name.c_str());
}


static void composite_inner()
{
    const std::string str =
	config::cfg.get_lowercase("InnerBoundary", "individual");

    composite_inner_name = str;
    if (str == "individual") {
	return;
    } else if (str == "zerogradient") {
	sigma_inner_name = "zerogradient";
	energy_inner_name = "zerogradient";
	vrad_inner_name = "zerogradient";
    } else if (str == "outflow") {
	sigma_inner_name = "zerogradient";
	energy_inner_name = "zerogradient";
	vrad_inner_name = "outflow";
    } else if (str == "viscous") {
	sigma_inner_name = "zerogradient";
	energy_inner_name = "zerogradient";
	vrad_inner_name = "viscous";
    } else if (str == "reflecting") {
	sigma_inner_name = "zerogradient";
	energy_inner_name = "zerogradient";
	vrad_inner_name = "reflecting";
    } else if (str == "reference") {
	sigma_inner_name = "reference";
	energy_inner_name = "reference";
	vrad_inner_name = "reference";
	} else if (str == "centerofmass") {
	sigma_inner_name = "none";
    energy_inner_name = "none";
    vrad_inner_name = "none";
    vaz_inner_name = "none";
    } else if (str == "custom") {
	sigma_inner_name = "none";
    energy_inner_name = "none";
    vrad_inner_name = "none";
    vaz_inner_name = "none";
	} else {
	throw std::runtime_error(
	    "Unknown boundary condition for inner boundary: " + str);
    }
    if (str != "individual") {
	logging::print_master(LOG_INFO "BC: Inner composite = %s\n",
			      composite_inner_name.c_str());
    }
}

static void composite_outer()
{
    const std::string str =
	config::cfg.get_lowercase("OuterBoundary", "individual");
    composite_outer_name = str;
    if (str == "individual") {
	return;
    } else if (str == "zerogradient") {
	sigma_outer_name = "zerogradient";
	energy_outer_name = "zerogradient";
	vrad_outer_name = "zerogradient";
    } else if (str == "outflow") {
	sigma_outer_name = "zerogradient";
	energy_outer_name = "zerogradient";
	vrad_outer_name = "outflow";
    } else if (str == "reflecting") {
	sigma_outer_name = "zerogradient";
	energy_outer_name = "zerogradient";
	vrad_outer_name = "reflecting";
    } else if (str == "reference") {
	sigma_outer_name = "reference";
	energy_outer_name = "reference";
	vrad_outer_name = "reference";
	} else if (str == "massoverflow") {
		throw std::logic_error("TODO: Selection of massoverflow not yet implemented.");
	} else if (str == "centerofmass") {
	sigma_outer_name = "none";
    energy_outer_name = "none";
    vrad_outer_name = "none";
    vaz_outer_name = "none";
    } else if (str == "custom") {
	sigma_outer_name = "none";
    energy_outer_name = "none";
    vrad_outer_name = "none";
    vaz_outer_name = "none";
	} else {
	throw std::runtime_error(
	    "Unknown boundary condition for outer boundary: " + str);
    }
    if (str != "individual") {
	logging::print_master(LOG_INFO "BC: Outer composite = %s\n",
			      composite_outer_name.c_str());
    }
}

void roche_lobe_overflow_setup_error_check(t_data &data){
    /*
     * Function must be called after 'parse_config()' of boundary conditions
     * has been called, which happens inside ReadVariables in main.cpp
     * and after 'data.get_planetary_system().init_system()' has been called,
     * which happens in main.cpp
     */


    if(!rochelobe_overflow){
	return;
    }

    const auto &nbody_sys = data.get_planetary_system();
    if(nbody_sys.get_number_of_planets() < rof_planet){
	die("Roche lobe overflowing companion %d not contained in Nbody system of size %d!",
	    rof_planet, nbody_sys.get_number_of_planets());
    }

    const t_planet &companion = nbody_sys.get_planet(rof_planet);

    if(RMAX > companion.get_semi_major_axis()){
	die("Roche lobe overflowing companion %s with a=%.3e must be outside of the simulation domain of size %.3e!",
	    companion.get_name().c_str(), companion.get_semi_major_axis(), RMAX);
    }

    if(companion.get_eccentricity() > 0.01){
	die("Roche lobe overflow only intended to be used with circular companions, but companion %s has an eccentricity of %.2g!",
	    companion.get_name().c_str(), companion.get_eccentricity());
    }

    const t_planet &primary = nbody_sys.get_planet(0);
    const double l1 = primary.get_dimensionless_roche_radius() *
			 companion.get_distance_to_primary();

    if(std::fabs(RMAX / l1 - 1) > 0.05){
	die("Lagrange l1 point %.3e of the primary must be equal to the outer domain radius %.3e to 5%% accuracy!",
	    l1, RMAX);
    }

}

void parse_config()
{

    composite_inner();
    composite_outer();

    sigma_inner();
    sigma_outer();

    energy_inner();
    energy_outer();

    vrad_inner();
    vrad_outer();

    vaz_inner();
    vaz_outer();

    viscous_outflow_speed = config::cfg.get<double>("ViscousOutflowSpeed", 1.0);


    // mass overflow
    rochelobe_overflow = config::cfg.get_flag("RocheLobeOverflow", "no");
	rof_variableTransfer = config::cfg.get_flag("ROFVariableTransfer", "no");
    rof_planet = config::cfg.get<int>("ROFplanet", 1);
    rof_temperature = config::cfg.get<double>("ROFtemperature", "1000.0 K", units::Temp0);
    rof_mdot = config::cfg.get<double>("ROFvalue", 10E-9, units::M0/units::T0);
    rof_rampingtime = config::cfg.get<double>("ROFrampingtime", 30.0);
	rof_averaging_time = config::cfg.get<double>("ROFaveragingtime", 10.0);
	rof_gamma = config::cfg.get<double>("ROFgamma", 0.5);

	damping_config();


    // check that damping in center of mass boundary does not conflict with regular damping
    if (composite_inner_name == "centerofmass") {
        for (unsigned int i = 0; i < damping_vector.size(); ++i) {
			if (damping_vector[i].type_inner != damping_none) {
                die("Damping is enabled twice. Here and through center of mass boundary. Set all inner damping variables to 'none'");
            }
		}
    }

    if (composite_outer_name == "centerofmass") {
        for (unsigned int i = 0; i < damping_vector.size(); ++i) {
            if (damping_vector[i].type_outer != damping_none) {
                die("Damping is enabled twice. Here and through center of mass boundary. Set all outer damping variables to 'none'");
            }
        }
    }

}


} // namespace boundary_conditions
